### PR TITLE
Add ability for field resolvers to modify the application context passed to nested field resolvers

### PR DESCRIPTION
**Pull request by hlship**
This change hits a few things:

- The selections pipeline has been modified using the parameters design pattern (a map of values, rather than the previous 3 parameters)
- RenderCommand protocol has been added: it defines a wrapper type that can modify the execution context
- ResolverResult is now about a single value, and there's a `with-error` wrapper function that wraps a resolved value with a RenderCommand
- The callback for `on-resolve!` now gets just a single value
- Logic in `executor` now applies any RenderCommands on the resolved value *before* passing it down the selector pipeline

... all that is to support:

- `with-context` wraps a resolved value such that nested field resolvers get a modified context

Also, I found that I can get reasonably accurate performance results using `quick-benchmark` rather than `benchmark`, and save a lot of time.

### CODE REVIEW

**Comment by hlship** [comment]
Closing until I see about reverting just the selector pipeline changes.

**Comment by hlship** [comment]
Ok, I made a few improvements and am now happy with it.  There's been a tiny uptick in execution time, but only in the very large queries (such as introspection).

**Comment by hlship** [comment]
Circle Ci is doing that thing again; tests do pass!

File: docs/resolve/context.rst [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+Here, the search term is provided to the ``products`` field, but is again needed by the ``highlighted_name``
+field, to highlight the parts of the name that match the search term.
+
+The resolver for the ``product`` field can communicate this "down tree" to the resolver for the ``highlighted_name`` field:
+
+.. literalinclude:: /_examples/mutable-context.edn
+   :language: clojure
+
+This new key, ``::search-term`` is only present in the context below ``products`` field.
```
Comment (candid82) [inline docs/resolve/context.rst]: Could this be solved by adding `::search-term` in `product` resolver's result value without modifying the context?

Comment (hlship) [inline docs/resolve/context.rst]: I put the `category` field in between them to justify the use of the context change.

Comment (candid82) [inline docs/resolve/context.rst]: Ah, I missed that part. Makes sense!

File: docs/resolve/context.rst [hunk 2]
```diff
@@ -0,0 +1,34 @@
+Application Context
+===================
+
+The application context passed to your field resolvers is normally set by the initial call to
+`com.walmartlabs.lacina/execute-query`.
+Lacinia uses the context for its own book-keeping (they keys it places into the map are namespaced to
```
Comment (sashton) [inline docs/resolve/context.rst]: typo: `they`

File: src/com/walmartlabs/lacinia/executor.clj [hunk 3]
```diff
@@ -44,14 +43,13 @@
...
-  each error, including location and arguments."
-  [field-selection error-maps]
-  (when (seq error-maps)
-    (let [enhanced-data (ex-info-map field-selection nil)]
-      (map
-        #(merge % enhanced-data)
-        error-maps))))
+  "From an error map, or a collection of error maps, add additional data to
+  each error, including location and arguments.  Returns a seq of error maps."
+  [field-selection error]
```
Comment (sashton) [inline src/com/walmartlabs/lacinia/executor.clj]: should `error` be renamed `errors` to match the function name?

File: docs/resolve/context.rst (hunk 2 above)
Comment (hlship) [inline docs/resolve/context.rst]: ok

File: src/com/walmartlabs/lacinia/executor.clj (hunk 3 above)
Comment (hlship) [inline src/com/walmartlabs/lacinia/executor.clj]: it's that darned `error-map-or-seq-of-error-maps` flexibility.

File: src/com/walmartlabs/lacinia/resolve.clj [hunk 4]
```diff
@@ -25,63 +61,79 @@
...
 
 (defn resolve-promise
   "Returns a ResolverResultPromise.
 
    A value must be resolved and ultimately provided via [[deliver!]]."
   []
-  (let [realized-result (promise)
-        callback-promise (promise)]
+  (let [*result (atom ::unset)
+        *callback (atom nil)]
```
Comment (sashton) [inline src/com/walmartlabs/lacinia/resolve.clj]: Why the change from `promise`s to `atom`s?

Comment (hlship) [inline src/com/walmartlabs/lacinia/resolve.clj]: I peeked inside the implementation of promise and found ... an atom.  Which surprised me. If I was really concerned about race conditions, probably use some java.util.concurrent stuff ... but the use case for all this is one thread perhaps creates the ResolverResultPromise and hands it off to another thread that delivers the value.  It would probably be nice to make this 100% bullet proof, but I don't see it as urgent.

File: test/com/walmartlabs/lacinia/resolver_result_test.clj [hunk 5]
```diff
@@ -12,32 +12,29 @@
 
 (deftest callback-is-invoked-with-value-and-errors
```
Comment (sashton) [inline test/com/walmartlabs/lacinia/resolver_result_test.clj]: I don't see the `-and-errors` part of this test anymore.

File: src/com/walmartlabs/lacinia/resolve.clj (hunk 4 above)
Comment (sashton) [inline src/com/walmartlabs/lacinia/resolve.clj]: In this case, the semantics of `promise/deliver/realized?` feel to me to match the use case in the code, and I find the promise version easier to read. Is there a performance improvement in the `atom` version?

Comment (hlship) [inline src/com/walmartlabs/lacinia/resolve.clj]: No, and I think you are right about the readability. I've moved it back towards using promises.

**Comment by hlship** [comment]
I believe all review issues have been addressed.
